src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.ClassifiedLineCache.cs

```diff
@@ -0,0 +1,186 @@
+        /// can manipulate it easily. For example, removing it from its current location (in O(1) time) and adding it to
+        /// the end (also O(1)).
+        /// </summary>
+        private readonly Dictionary<Span, LinkedListNode<SpanAndClassifiedSpans>> _spanToLruNode = [];
+
+        private void ClearIfDifferentSnapshot(ITextSnapshot snapshot)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            if (_snapshot != snapshot)
+            {
+                _lruList.Clear();
+                _spanToLruNode.Clear();
+                _snapshot = snapshot;
+            }
+        }
+
+        public bool TryUseCache(SnapshotSpan snapshotSpan, SegmentedList<ClassifiedSpan> classifications)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            ClearIfDifferentSnapshot(snapshotSpan.Snapshot);
+
+            if (!_spanToLruNode.TryGetValue(snapshotSpan.Span, out var node))
+                return false;
+
+            // Move this node to the front of the LRU list.
+            _lruList.Remove(node);
+            _lruList.AddLast(node);
+
+            // AddRange is optimized to take a SegmentedList and copy directly from it into the result list.
+            classifications.AddRange(node.Value.ClassifiedSpans);
+            return true;
+        }
+
+        public void Update(SnapshotSpan snapshotSpan, SegmentedList<ClassifiedSpan> newClassifications)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            if (newClassifications.Count > MaxClassificationsCount)
+                return;
+
+            // Clear out cached data if we've moved to a different snapshot.
+            var span = snapshotSpan.Span;
+            ClearIfDifferentSnapshot(snapshotSpan.Snapshot);
+
+            if (_spanToLruNode.TryGetValue(span, out var existingNode))
+            {
+                UpdateExistingEntryInCache(newClassifications, existingNode);
+            }
+            else
+            {
+                AddNewEntryToCache(span, newClassifications);
+            }
+
+            Contract.ThrowIfTrue(_lruList.Count > CacheSize);
+            Contract.ThrowIfTrue(_lruList.Count != _spanToLruNode.Count);
+        }
+
+        /// <summary>
+        /// Helper that allows us to reuse the classified spans list inside <paramref name="existingNode"/> list,
+        /// updating it to have all the classifications in <paramref name="classifications"/>.
+        /// </summary>
+        private static void UpdateLruNodeClassifications(
+            LinkedListNode<SpanAndClassifiedSpans> existingNode,
+            SegmentedList<ClassifiedSpan> classifications)
+        {
+            existingNode.Value.ClassifiedSpans.Clear();
+
+            // AddRange is optimized to take a SegmentedList and copy directly from it into the result list.
+            existingNode.Value.ClassifiedSpans.AddRange(classifications);
+        }
+
+        private void UpdateExistingEntryInCache(
```

Comment:
>UpdateExistingEntryInCache [](http://example.com/codeflow?start=21&length=26)

Can you clarify for me when this is currently called? Same snapshot, same span, but we want to specify new values?

Reply:
updated ot remove this.